tempodb/encoding/backend_block.go

```diff
@@ -139,3 +148,126 @@ func (b *BackendBlock) NewIndexReader() (common.IndexReader, error) {
 func (b *BackendBlock) BlockMeta() *backend.BlockMeta {
 	return b.meta
 }
+
+func (b *BackendBlock) FindTraceByID(ctx context.Context, id common.ID) (*tempopb.Trace, error) {
+	obj, err := b.find(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	if obj == nil {
+		// Not found in this block
+		return nil, nil
+	}
+
+	dec, err := model.NewObjectDecoder(b.meta.DataEncoding)
+	if err != nil {
+		return nil, err
+	}
+	return dec.PrepareForRead(obj)
+}
+
+func (b *BackendBlock) Search(ctx context.Context, req *tempopb.SearchRequest, opts ...SearchOption) (resp *tempopb.SearchResponse, err error) {
+
+	// Options
+	opt := SearchOptions{
+		chunkSizeBytes: 10_000_000,
+	}
+	for _, o := range opts {
+		o(&opt)
+	}
+
+	decoder, err := model.NewObjectDecoder(b.meta.DataEncoding)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create NewDecoder: %w", err)
+	}
+
+	// Iterator
+	var iter Iterator
+	if opt.totalPages > 0 {
+		iter, err = b.partialIterator(opt.chunkSizeBytes, opt.startPage, opt.totalPages)
+	} else {
+		iter, err = b.Iterator(opt.chunkSizeBytes)
+	}
+	if err != nil {
+		return nil, err
+	}
+	if opt.prefetchTraceCount > 0 {
+		iter = NewPrefetchIterator(ctx, iter, opt.prefetchTraceCount)
+	}
+	defer iter.Close()
+
+	respMtx := sync.Mutex{}
+	resp = &tempopb.SearchResponse{
+		Metrics: &tempopb.SearchMetrics{},
+	}
+
+	var searchErr error
+	wg := boundedwaitgroup.New(5)
```

Comment:
I'm concerned about this b/c it will impact the amount of resources needed by the serverless functions which do not concurrently iterate through the block. it would be interesting to see if this ups throughput but I would like to do a test before merging this.

my instinct is to remove this concurrency and require the caller to do more in parallel.

Reply:
If we made this a parameter and serverless passed `1`, would that negatively impact it also?   Considering what a boundedwaitgroup of 1 looks like, seems so and inefficient.  Will remove.   

The remaining impacted caller is the querier for backend search, when it prefers self for a search shard. Since there's already good parallelization over the whole request, thinking it's ok as-is if an individual shard goes from concurrency=5 to 1. Thoughts?

Reply:
Agree. Let's just remove the boundedwaitgroup/concurrency here and we can increase `search_prefer_self` if we want to increase the amount of work a querier will do.